### PR TITLE
#161776762 Edit articles

### PR DESCRIPTION
**Pull request by samuelbwambale**
### What does this PR do?

Enables users to edit articles which they authored

#### Description of Task to be completed?

Currently, the users are unable to update articles. 

This feature enables users to edit articles. Users can only edit the articles they authored.

#### How should this be manually tested?

Pull the changes on a new branch and run `npm install`
run `npm start` to start the server
Log into the app using a registered email and password
Click on profile and select `New Article` to add a new article
Select the created article from the list of articles by clicking on `Read more` link
On loading the article, click on edit button and change any of the fields
Click on save button and see the updated article on the profile

#### What are the relevant pivotal tracker stories?

- [#161776762](https://www.pivotaltracker.com/story/show/161776762)

## Checklist: I completed these to help reviewers

- [x] My IDE is configured to follow the [**ESlint Conventions**](https://github.com/eslint/eslint)

- [x] I have **added tests** to cover my changes.

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the develop branch.

### CODE REVIEW

**Review by RonKbS — COMMENTED** [review]
Nice work @samuelbwambale. 
There is a bug that should be easily fixed:
When I click on `Read more` it shows the article briefly before returning to the dashboard

File: src/actions/types.js [hunk 1]
```diff
@@ -7,8 +7,17 @@ export const SOCIAL_LOGIN_SUCCESS = 'SOCIAL_LOGIN_SUCCESS';
 export const GET_PROFILE_PAYLOAD = 'GET_PROFILE_PAYLOAD';
 export const GET_PROFILE_INITIATED = 'GET_PROFILE_INITIATED';
 
-export const GET_ALL_ARTICLES_SUCCESS = 'GET_ALL_ARTICLES_SUCCESS';
 export const CREATE_ARTICLE_SUCCESS = 'CREATE_ARTICLE_SUCCESS';
 export const CREATE_ARTICLE_INITIATED = 'CREATE_ARTICLE_INITIATED';
 export const CREATE_ARTICLE_ERROR = 'CREATE_ARTICLE_ERROR';
 export const LOGOUT_USER = 'LOGOUT_USER';
+export const SEND_RESET_LINK_SUCCESS = 'SEND_RESET_LINK_SUCCESS';
```
Comment (RonKbS) [inline src/actions/types.js]: Please remove these reset password changes

File: src/actions/articleActions.js [hunk 2]
```diff
@@ -34,3 +32,47 @@ export const postArticle = postData => dispatch => {
...
+  toast.dismiss();
+  return axiosInstance
+    .put(`/api/articles/${slug}/`, newData)
+    .then(response => {
+      dispatch(editArticleSuccess(true));
+      toast.success(
+        response.data.message,
+        { autoClose: 3500, hideProgressBar: true },
+        {
+          position: toast.POSITION.TOP_CENTER,
```
Comment (MuhanguziDavid) [inline src/actions/articleActions.js]: @samuelbwambale , please remove this line, the toast position was already set.

File: src/components/Articles/ViewArticle.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    this.setState({ description });
+
+    if (nextProps.editArticleSuccess === true) {
+      this.setState({ isEditing: false });
+      const { history } = this.props;
+      history.push('/dashboard');
+    }
+  }
+
+
```
Comment (MuhanguziDavid) [inline src/components/Articles/ViewArticle.js]: Please remove redundant line

File: src/components/Articles/ViewArticle.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  toggleEdit() {
+    const { isEditing } = this.state;
+    this.setState({ isEditing: !isEditing });
+  }
+
+  render() {
+    const { title, description, body } = this.state;
+    const { loading } = this.props;
+    const { articlePayload, articlePayload: { article } } = this.props;
+    const { isEditing } = this.state;
```
Comment (MuhanguziDavid) [inline src/components/Articles/ViewArticle.js]: Lines 92 and 95 can be put on the same line,
i.e. `const { title, description, body, isEditing } = this.state;`

Lines 93 and 94 can also be put on the same line.

File: src/tests/actions/articleActions.test.js [hunk 5]
```diff
@@ -66,4 +96,65 @@ describe('articleActions', () => {
...
+    await flushAllPromises();
+    expect(store.getActions()).toEqual(
+      [
+        { type: EDIT_ARTICLE_INITIATED, payload: true },
+        { type: EDIT_ARTICLE_SUCCESS, payload: true },
+      ],
+    );
+  });
+
+  // it('should not update an article if not the author', async () => {
```
Comment (MuhanguziDavid) [inline src/tests/actions/articleActions.test.js]: Please remove the commented block of code

File: src/actions/types.js (hunk 1 above)
Comment (samuelbwambale) [reply src/actions/types.js]: @RonKbS Reset-password changes have been removed after rebasing develop branch

File: src/actions/articleActions.js (hunk 2 above)
Comment (samuelbwambale) [reply src/actions/articleActions.js]: @MuhanguziDavid Line removed to take the default toast settings

File: src/components/Articles/ViewArticle.js (hunk 4 above)
Comment (samuelbwambale) [reply src/components/Articles/ViewArticle.js]: @MuhanguziDavid code refactored as advised

File: src/tests/actions/articleActions.test.js (hunk 5 above)
Comment (samuelbwambale) [reply src/tests/actions/articleActions.test.js]: @MuhanguziDavid Commented code has been removed from the PR

**Comment by samuelbwambale** [comment]
@RonKbS the bug causing an article not to be loaded was fixed

File: src/actions/types.js (hunk 1 above)
Comment (RonKbS) [reply src/actions/types.js]: @samuelbwambale these particular action_types should also be removed since the updated PR for them has not yet been merged.

File: src/actions/articleActions.js [hunk 6]
```diff
@@ -113,3 +115,28 @@ export const likeDislike = (payload, slug) => dispatch => {
...
+        { autoClose: 3500, hideProgressBar: true },
+      );
+    })
+    .catch((error) => {
+      let errorMessage = '';
+      if (error.response.status === 403) {
+        errorMessage = 'Re-login and try again';
+      }
+      if (error.response.status === 404) {
+        errorMessage = 'Please enter valid data';
```
Comment (RonKbS) [inline src/actions/articleActions.js]: @samuelbwambale I think we may need a more user-friendly message here. Developers might feel at home when they see this message but the same probably can't be said for other types of authors.

File: src/tests/actions/articleActions.test.js [hunk 7]
```diff
@@ -185,6 +188,13 @@ describe('likeDislikeAction', () => {
   localStorage.setItem('auth_token', 'token');
   const payload = { reaction: 'Like' };
   const slug = 'testing-1-2-3';
+  const article_data = {
+    article: {
+      body: '<p>The first you’ll want tr, they may have already done this, so  components!</p>',
```
Comment (RonKbS) [inline src/tests/actions/articleActions.test.js]: There are some spelling mistakes here.

File: src/components/Articles/ViewArticle.js (hunk 3 above)
Comment (samuelbwambale) [reply src/components/Articles/ViewArticle.js]: @MuhanguziDavid Line removed

File: src/actions/types.js (hunk 1 above)
Comment (samuelbwambale) [reply src/actions/types.js]: @RonKbS The action_types have been removed

File: src/actions/articleActions.js (hunk 6 above)
Comment (samuelbwambale) [reply src/actions/articleActions.js]: I have updated the response to `Please enter a valid text`. Let me know if this suffices

File: src/tests/actions/articleActions.test.js (hunk 7 above)
Comment (samuelbwambale) [reply src/tests/actions/articleActions.test.js]: @RonKbS This was random text but has been updated to correct spellings

File: src/tests/components/NewArticleForm.test.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    wrapper.find('#add-article-form').simulate('submit');
+    expect(handleSubmit).toBeCalled();
+  });
+
+  it('should call handleChange when the there is an input in title/description', () => {
+    wrapper.find('#title').simulate('change');
+    expect(handleChange).toBeCalled();
+  });
+
+  it('should call handleEditorChange when the there is an input in text editor', () => {
```
Comment (kakaemma) [inline src/tests/components/NewArticleForm.test.js]: Consider changing `when the there` on line 47 and 52 to make the test more meaningful.

File: src/components/Articles/NewArticleForm.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
+import React from 'react';
+import Loader from 'react-loader';
+import PropTypes from 'prop-types';
+import ReactQuill from 'react-quill';
+import 'react-quill/dist/quill.snow.css';
+
+
```
Comment (kakaemma) [inline src/components/Articles/NewArticleForm.js]: Kindly consider removing this redundant line.

File: src/tests/components/NewArticleForm.test.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      resetForm={resetForm}
+      handleChange={handleChange}
+      handleEditorChange={handleEditorChange}
+      body={body}
+      title={title}
+      description={description}
+      loading={loading}
+    />);
+  });
+
```
Comment (kakaemma) [inline src/tests/components/NewArticleForm.test.js]: This line too should be removed and all the other redundant lines.

File: src/components/Articles/ViewArticle.js [hunk 11]
```diff
@@ -1,41 +1,157 @@
...
+
+  resetForm = () => {
+    this.setState({
+      title: '',
+      description: '',
+      body: '',
+    });
+  };
+
+  toggleEdit() {
```
Comment (kakaemma) [inline src/components/Articles/ViewArticle.js]: Consider making this an arrow function which binds automatically to the class. Then remove line `21`.

File: src/tests/components/NewArticleForm.test.js (hunk 8 above)
Comment (samuelbwambale) [reply src/tests/components/NewArticleForm.test.js]: @kakaemma lines 47 and 52 have been updated with more descriptive comment

File: src/components/Articles/NewArticleForm.js (hunk 9 above)
Comment (samuelbwambale) [reply src/components/Articles/NewArticleForm.js]: @kakaemma Redundant line has been removed

File: src/tests/components/NewArticleForm.test.js (hunk 10 above)
Comment (samuelbwambale) [reply src/tests/components/NewArticleForm.test.js]: @kakaemma Redundant line 31 has been removed. Please note that there are some other blank lines left for easy readability of code.

File: src/components/Articles/ViewArticle.js (hunk 11 above)
Comment (samuelbwambale) [reply src/components/Articles/ViewArticle.js]: Thanks, @kakaemma `toggleEdit()` has been redefined as an arrow function and line `21` deleted.

**Review by MuhanguziDavid — APPROVED** [review]
LGTM

**Review by mendozabree — APPROVED** [review]
@samuelbwambale I was able to edit an article!
LGTM

**Review by RonKbS — APPROVED** [review]
LGTM

File: src/actions/articleActions.js [hunk 12]
```diff
@@ -62,16 +65,15 @@ export const fetchComments = article => dispatch => {
 
 export const fetchArticles = () => dispatch => {
   dispatch(getArticlesInitiated(true));
-  return axiosInstance
```
Comment (dondrzzy) [inline src/actions/articleActions.js]: Why did you remove the return statement?

File: src/actions/articleActions.js [hunk 13]
```diff
@@ -62,16 +65,15 @@ export const fetchComments = article => dispatch => {
...
+  axiosInstance
     .get('/api/article/')
     .then((response) => {
       dispatch(getAllArticles(response.data.article));
     });
 };
 
 export const fetchSpecificArticle = slug => dispatch => {
-  dispatch(getArticlesInitiated(true));
-  return axiosInstance
```
Comment (dondrzzy) [inline src/actions/articleActions.js]: Why did you remove both lines of code?

File: src/actions/articleActions.js [hunk 14]
```diff
@@ -113,3 +115,28 @@ export const likeDislike = (payload, slug) => dispatch => {
       return toast.error(error.response.data.detail, { autoClose: false, hideProgressBar: true });
     });
 };
+
+export const updateArticle = (slug, newData) => dispatch => {
+  toast.dismiss();
+  dispatch(editArticleInititated(true));
+  axiosInstance
```
Comment (dondrzzy) [inline src/actions/articleActions.js]: Add a return statement. `updateArticle` should return a promise.

File: src/actions/articleActions.js [hunk 15]
```diff
@@ -113,3 +115,28 @@ export const likeDislike = (payload, slug) => dispatch => {
...
+    .catch((error) => {
+      let errorMessage = '';
+      if (error.response.status === 403) {
+        errorMessage = 'Re-login and try again';
+      }
+      if (error.response.status === 404) {
+        errorMessage = 'Please enter valid text';
+      }
+      dispatch(editArticleError(errorMessage));
+      toast.error(errorMessage, { autoClose: false, hideProgressBar: true });
```
Comment (dondrzzy) [inline src/actions/articleActions.js:155]: If you are notifying the user of this error, why add it to state?

File: src/components/Articles/ViewArticle.js [hunk 16]
```diff
@@ -1,41 +1,156 @@
...
   }
 
-  render() {
-    const { articlePayload, articlePayload: { article } } = this.props;
+  componentWillReceiveProps(nextProps) {
+    const { articlePayload } = nextProps;
+    const { article: { title, body, description } } = articlePayload;
+    this.setState({ title });
+    this.setState({ body });
+    this.setState({ description });
```
Comment (dondrzzy) [inline src/components/Articles/ViewArticle.js]: Use one `setState` statement.

File: src/tests/components/NewArticle.test.js [hunk 17]
```diff
@@ -25,39 +28,39 @@ describe('NewArticle component', () => {
 
   beforeEach(() => {
     mockStore({});
-    wrapper = shallow(<NewArticle {...props} />);
+    wrapper = shallow(<NewArticle
+      {...props}
+      addArticle={addArticle}
+      resetForm={resetForm}
+      handleEditorChange={handleEditorChange}
+      handleChange={handleChange}
```
Comment (dondrzzy) [inline src/tests/components/NewArticle.test.js]: Add all these props to the props const.

File: src/tests/components/NewArticleForm.test.js [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+  const mockStore = configureMockStore();
+  let wrapper;
+  const handleSubmit = jest.fn();
+  const resetForm = jest.fn();
+  const handleChange = jest.fn();
+  const handleEditorChange = jest.fn();
+  const title = '';
+  const description = '';
+  const body = '';
+  const loading = false;
```
Comment (dondrzzy) [inline src/tests/components/NewArticleForm.test.js]: Use a props object for all these props and pass its spread.

File: src/tests/components/NewArticleForm.test.js [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+  });
+
+  it('should call resetForm when the clear button is clicked', () => {
+    wrapper.find('#clear-button').simulate('click', { preventDefault() {} });
+    expect(resetForm).toBeCalled();
+  });
+
+  it('should call handleSubmit when the form is submitted', () => {
+    wrapper.find('#add-article-form').simulate('submit');
+    expect(handleSubmit).toBeCalled();
```
Comment (dondrzzy) [inline src/tests/components/NewArticleForm.test.js]: Add a test showing what describe `handleSubmit` does. Same goes for the rest

File: src/tests/components/NewArticleForm.test.js [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+    />);
+  });
+
+  it('should render correctly', () => {
+    expect(wrapper).toMatchSnapshot();
+  });
+
+  it('should call resetForm when the clear button is clicked', () => {
+    wrapper.find('#clear-button').simulate('click', { preventDefault() {} });
+    expect(resetForm).toBeCalled();
```
Comment (dondrzzy) [inline src/tests/components/NewArticleForm.test.js]: Add a test showing what describe `resetForm ` does.

**Review by dondrzzy — COMMENTED** [review]
Please refactor

File: src/actions/articleActions.js (hunk 12 above)
Comment (samuelbwambale) [reply src/actions/articleActions.js]: The return statement has been replaced.

File: src/actions/articleActions.js (hunk 13 above)
Comment (samuelbwambale) [reply src/actions/articleActions.js]: The `getArticlesInitiated(true)` was not adding anything to the app. Instead it was disabling the loading of the list of articles. This was removed. Without any other statement in the arrow function, ESLint suggests that the return statement be removed

File: src/actions/articleActions.js (hunk 14 above)
Comment (samuelbwambale) [reply src/actions/articleActions.js]: Return statement has been added to return the promise

File: src/actions/articleActions.js (hunk 15 above)
Comment (samuelbwambale) [reply src/actions/articleActions.js:155]: This dispatch handles change in state for other state variables such as `loading` besides adding the error to the state

File: src/components/Articles/ViewArticle.js (hunk 16 above)
Comment (samuelbwambale) [reply src/components/Articles/ViewArticle.js]: This piece of code has been refactored as advised

File: src/tests/components/NewArticle.test.js (hunk 17 above)
Comment (samuelbwambale) [reply src/tests/components/NewArticle.test.js]: The constants have been added into a single object `const props {}`

File: src/tests/components/NewArticleForm.test.js (hunk 19 above)
Comment (samuelbwambale) [reply src/tests/components/NewArticleForm.test.js]: This was handled in the file `NewArticle.test.js` which has the tests for the main component

File: src/tests/components/NewArticleForm.test.js (hunk 20 above)
Comment (samuelbwambale) [reply src/tests/components/NewArticleForm.test.js]: This was handled in the file `NewArticle.test.js` which has the tests for the main component

File: src/tests/components/NewArticleForm.test.js (hunk 18 above)
Comment (samuelbwambale) [reply src/tests/components/NewArticleForm.test.js]: The constants have been added into a single object const props {}

File: src/actions/articleActions.js (hunk 13 above)
Comment (dondrzzy) [reply src/actions/articleActions.js]: When you fetch a new article and it takes longer, what does the user see while that happens?

File: src/components/Articles/EditArticle.js [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+      match: {
+        params: { slug },
+      },
+    } = this.props;
+    fetchSpecificArticle(slug);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { articlePayload } = nextProps;
+    if (nextProps.articlePayload !== this.props.articlePayload) {
```
Comment (dondrzzy) [inline src/components/Articles/EditArticle.js]: Did you intentionally ignore this linting error?

File: src/components/Articles/EditArticle.js [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+    const {
+      title,
+      description,
+      body,
+      isEditing,
+    } = this.state;
+    const { loading, articlePayload, articlePayload: { article } } = this.props;
+    if (isEditing) {
+      return (
+        <CreateArticleForm
```
Comment (dondrzzy) [inline src/components/Articles/EditArticle.js]: Rename this to `ArticleForm`. `Create` does not sound right in `Edit`.

File: src/actions/articleActions.js [hunk 23]
```diff
@@ -113,3 +117,28 @@ export const likeDislike = (payload, slug) => dispatch => {
...
+        { autoClose: 3500, hideProgressBar: true },
+      );
+    })
+    .catch((error) => {
+      let errorMessage = '';
+      if (error.response.status === 403) {
+        errorMessage = 'Re-login and try again';
+      }
+      if (error.response.status === 404) {
+        errorMessage = 'Please enter valid text';
```
Comment (dondrzzy) [inline src/actions/articleActions.js]: Does `valid text` help the user determine whether its the `title, body or description` with the issue? Is there an error message that returns with this error response?

File: src/tests/reducers/articleReducer.test.js [hunk 24]
```diff
@@ -52,10 +57,32 @@ describe('articlesReducer', () => {
...
   });
 
-  it('should add CREATE_ARTICLE_SUCCESS to true when an article is posted successfully', () => {
+  it('should set the loader when GET_ALL_ARTICLES_INITIATED is true', () => {
+    const action = {
+      type: GET_ALL_ARTICLES_INITIATED,
+      payload: true,
+    };
+    const currentState = articlesReducer(initialState, action);
+    expect(currentState).toEqual({
```
Comment (dondrzzy) [inline src/tests/reducers/articleReducer.test.js:64]: Everyone seems to be doing this which is making this file ambiguous. How about this:

expect(...).toEqual({
    ...initialState,
    loading: action.payload,
});
@samuelbwambale . If it works, advise the next victims.

**Review by dondrzzy — COMMENTED** [review]
Please refactor

File: src/actions/articleActions.js (hunk 13 above)
Comment (samuelbwambale) [reply src/actions/articleActions.js]: I added a loader to show as the article is being fetched

File: src/components/Articles/EditArticle.js (hunk 22 above)
Comment (samuelbwambale) [reply src/components/Articles/EditArticle.js]: `CreateArticleForm` has been renamed to `ArticleForm`

File: src/tests/reducers/articleReducer.test.js (hunk 24 above)
Comment (samuelbwambale) [reply src/tests/reducers/articleReducer.test.js:64]: I have refactored this and shared with the team

File: src/actions/articleActions.js (hunk 23 above)
Comment (samuelbwambale) [reply src/actions/articleActions.js]: The front-end already caters for validation of `title` and `description` fields. I have updated the message to point to the article body specifically.

File: src/components/Articles/EditArticle.js (hunk 21 above)
Comment (samuelbwambale) [reply src/components/Articles/EditArticle.js]: This was skipped accidentally. It has been restructured.
